In [1]:
import os
import sys


repo_dir = os.path.dirname(os.getcwd())


output_dir = os.path.join(repo_dir, "output")

sys.path.append(repo_dir)

In [2]:
from collections import Counter
import numpy as np
import pandas as pd
import plotly.graph_objects as go

from data_utils import read_cmu_mosei, create_dataset

/tmp/ipykernel_26946/3810907228.py:3: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


# 1. Load data

In [3]:
words, labels = read_cmu_mosei(
    data_dir="/home/vinicius/Documentos/Repositories/msc/data"
)

train, validation, test = create_dataset(words, labels)

[2024-01-30 19:45:48.261] | Success | Computational sequence read from file /home/vinicius/Documentos/Repositories/msc/data/CMU_MOSEI_TimestampedPhones.csd ...
[2024-01-30 19:45:48.426] | Status  | Checking the integrity of the <phoneme> computational sequence ...
[2024-01-30 19:45:48.427] | Status  | Checking the format of the data in <phoneme> computational sequence ...


[2024-01-30 19:45:50.483] | Success | <phoneme> computational sequence data in correct format.
[2024-01-30 19:45:50.483] | Status  | Checking the format of the metadata in <phoneme> computational sequence ...
[2024-01-30 19:45:50.483] | Warning | <phoneme> computational sequence does not have all the required metadata ... continuing 
[2024-01-30 19:45:50.484] | Success | Computational sequence read from file /home/vinicius/Documentos/Repositories/msc/data/CMU_MOSEI_TimestampedWordVectors.csd ...
[2024-01-30 19:45:50.601] | Status  | Checking the integrity of the <glove_vectors> computational sequence ...
[2024-01-30 19:45:50.601] | Status  | Checking the format of the data in <glove_vectors> computational sequence ...


[2024-01-30 19:45:52.469] | Success | <glove_vectors> computational sequence data in correct format.
[2024-01-30 19:45:52.469] | Status  | Checking the format of the metadata in <glove_vectors> computational sequence ...
[2024-01-30 19:45:52.469] | Warning | <glove_vectors> computational sequence does not have all the required metadata ... continuing 
[2024-01-30 19:45:52.470] | Success | Computational sequence read from file /home/vinicius/Documentos/Repositories/msc/data/CMU_MOSEI_Labels.csd ...
[2024-01-30 19:45:52.566] | Status  | Checking the integrity of the <All Labels> computational sequence ...
[2024-01-30 19:45:52.566] | Status  | Checking the format of the data in <All Labels> computational sequence ...


[2024-01-30 19:45:54.205] | Success | <All Labels> computational sequence data in correct format.
[2024-01-30 19:45:54.205] | Status  | Checking the format of the metadata in <All Labels> computational sequence ...
[2024-01-30 19:45:54.205] | Warning | <All Labels> computational sequence does not have all the required metadata ... continuing 
[2024-01-30 19:45:54.206] | Success | Computational sequence read from file /home/vinicius/Documentos/Repositories/msc/data/CMU_MOSEI_TimestampedWords.csd ...
[2024-01-30 19:45:54.361] | Status  | Checking the integrity of the <words> computational sequence ...
[2024-01-30 19:45:54.361] | Status  | Checking the format of the data in <words> computational sequence ...


[2024-01-30 19:45:56.166] | Success | <words> computational sequence data in correct format.
[2024-01-30 19:45:56.166] | Status  | Checking the format of the metadata in <words> computational sequence ...
[2024-01-30 19:45:56.166] | Warning | <words> computational sequence does not have all the required metadata ... continuing 
[2024-01-30 19:45:56.166] | Success | Dataset initialized successfully ... 


In [4]:
# Avaliação do tamanho do dataset
train_senteces = train.shape[0]
validation_sentences = validation.shape[0]
test_sentences = test.shape[0]
print(f"Número de sentenças de treinamento: {train_senteces}")
print(f"Número de sentenças do validação:   {validation_sentences}")
print(f"Número de sentenças do teste:       {test_sentences}")
print(
    f"Número total de sentenças:          {train_senteces + validation_sentences + test_sentences}"
)

df = pd.concat([train, validation, test], ignore_index=True)
display(df)

Número de sentenças de treinamento: 15807
Número de sentenças do validação:   1818
Número de sentenças do teste:       4512
Número total de sentenças:          22137


Sentence  Sentiment  \
0      it is written that we should not be yoked with...  -1.000000   
1      if we are to be children of god we have to be ...  -0.333333   
2      this is a very difficult message because even ...  -1.333333   
3      and even in your own family or your own friend...  -0.333333   
4      but what i do know is if you ask for the holy ...   1.333333   
...                                                  ...        ...   
22132  a lot of the movie is just more awkward than i...  -2.000000   
22133  suggest going out there and finding save yourself  -2.000000   
22134  the money get yourself another nicholas cage m...  -2.666667   
22135                  i don't know like face off that's   0.000000   
22136     a good john woo nicholas cage movie watch that   1.000000   

       Happiness   Sadness     Anger      Fear   Disgust  Surprise  
0       0.000000  0.333333  2.000000  0.000000  0.000000  0.333333  
1       0.000000  0.000000  0.333333  0.000000  0.666667  0.000000  
2       0.000000  0.333333  1.000000  0.333333  0.000000  0.000000  
3       0.000000  0.000000  0.000000  0.000000  0.000000  0.333333  
4       0.333333  0.000000  0.000000  0.000000  0.000000  0.000000  
...          ...       ...       ...       ...       ...       ...  
22132   0.000000  0.333333  1.000000  0.000000  1.666667  0.000000  
22133   0.000000  0.666667  0.333333  0.000000  0.333333  0.000000  
22134   0.000000  0.333333  0.666667  0.000000  0.666667  0.000000  
22135   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  
22136   0.000000  0.000000  0.333333  0.000000  0.000000  0.000000  

[22137 rows x 8 columns]

# 2. Exploration Data Analysis

## 2.1 Neutral Sentences

In [5]:
# Decisão de classificar as sentenças como neutras
def neutral_label(row):
    emotions = ["Happiness", "Sadness", "Anger", "Fear", "Disgust", "Surprise"]

    if all(row[emotion] == 0 for emotion in emotions):
        return "Neutral"
    else:
        return "Non-Neutral"


df["Emotion_Label"] = df.apply(neutral_label, axis=1)
df.loc[df["Emotion_Label"] == "Neutral"]

Sentence  Sentiment  \
29     that right message at the right time from stra...   0.666667   
49     uhh a lot of the like you know a lot of the li...  -0.333333   
63             definitely keep this for the grownups uhh   0.666667   
71                   it where i know where it's gonna be   0.666667   
82     well russia didn't have the resources we have ...   0.000000   
...                                                  ...        ...   
22079  and uhh it's just like i guess like stutter sp...   0.000000   
22110  so india can clearly be a superpower it's a qu...   0.000000   
22115  here's a clip clip playing on the television s...   0.000000   
22124  second you know what i mean like throwing a th...  -0.333333   
22135                  i don't know like face off that's   0.000000   

       Happiness  Sadness  Anger  Fear  Disgust  Surprise Emotion_Label  
29           0.0      0.0    0.0   0.0      0.0       0.0       Neutral  
49           0.0      0.0    0.0   0.0      0.0       0.0       Neutral  
63           0.0      0.0    0.0   0.0      0.0       0.0       Neutral  
71           0.0      0.0    0.0   0.0      0.0       0.0       Neutral  
82           0.0      0.0    0.0   0.0      0.0       0.0       Neutral  
...          ...      ...    ...   ...      ...       ...           ...  
22079        0.0      0.0    0.0   0.0      0.0       0.0       Neutral  
22110        0.0      0.0    0.0   0.0      0.0       0.0       Neutral  
22115        0.0      0.0    0.0   0.0      0.0       0.0       Neutral  
22124        0.0      0.0    0.0   0.0      0.0       0.0       Neutral  
22135        0.0      0.0    0.0   0.0      0.0       0.0       Neutral  

[2974 rows x 9 columns]

In [6]:
# Existem frases classificadas sem nenhuma emocao?
neutral_phrases = df.loc[df["Emotion_Label"] == "Neutral"]

neutral_phrases.describe().loc[["min", "25%", "50%", "75%", "max"]]

Sentiment  Happiness  Sadness  Anger  Fear  Disgust  Surprise
min  -2.333333        0.0      0.0    0.0   0.0      0.0       0.0
25%   0.000000        0.0      0.0    0.0   0.0      0.0       0.0
50%   0.000000        0.0      0.0    0.0   0.0      0.0       0.0
75%   0.333333        0.0      0.0    0.0   0.0      0.0       0.0
max   2.666667        0.0      0.0    0.0   0.0      0.0       0.0

### 2.1.1 Percentage of Neutral Sentences in dataset

In [7]:
count_neutral_phrases = neutral_phrases.shape[0]

count_df = df.shape[0]

porcentage_neutral_phrases = np.round(count_neutral_phrases / count_df, 2) * 100


print(f"Porcentage of neutral phrases: {porcentage_neutral_phrases} %")

Porcentage of neutral phrases: 13.0 %


### 2.1.2 Quantity of neutral sentences per sentiment

In [8]:
count = neutral_phrases.groupby(by="Sentiment").count()["Sentence"]

percentage = neutral_phrases["Sentiment"].value_counts(normalize=True) * 100

summary = pd.DataFrame({"Count": count, "Percentage": percentage})

summary

Count  Percentage
Sentiment                   
-2.333333      1    0.033625
-2.000000      9    0.302623
-1.666667     38    1.277740
-1.333333     49    1.647613
-1.000000     57    1.916611
-0.666667    134    4.505716
-0.333333    247    8.305313
-0.166667      1    0.033625
 0.000000   1395   46.906523
 0.166667      1    0.033625
 0.333333    501   16.845999
 0.666667    285    9.583053
 1.000000    125    4.203093
 1.333333     75    2.521856
 1.500000      1    0.033625
 1.666667     33    1.109617
 2.000000     18    0.605245
 2.333333      3    0.100874
 2.666667      1    0.033625

### 2.1.3 Examples of Neutral Sentences with Positive Sentiment

In [9]:
positive_neutral_phrases = neutral_phrases.loc[neutral_phrases["Sentiment"] >= 2]

In [10]:
# Example of positive sentences
for _, s in positive_neutral_phrases.iterrows():
    sentence = s["Sentence"]
    width = 100
    print("-" * width)
    for i in range(0, len(sentence), width):
        print(sentence[i : i + width])
    print("-" * width)
    print()

----------------------------------------------------------------------------------------------------
i really enjoy watching will ferrell in this movie umm
----------------------------------------------------------------------------------------------------

----------------------------------------------------------------------------------------------------
they do y'all do a great job you're
----------------------------------------------------------------------------------------------------

----------------------------------------------------------------------------------------------------
for the past years the good people of jackson office have done their the words of martin luther bend
 the arc of arc of the moral justice to the employees of the jackson and you for your service to the
 and to the citizens of mississippi
----------------------------------------------------------------------------------------------------

--------------------------------------------------------------

### 2.1.4 Examples of Neutral Sentences with Negative Sentiment

In [11]:
negative_neutral_phrases = neutral_phrases.loc[neutral_phrases["Sentiment"] <= -2]

In [12]:
# Example of negative sentences
for _, s in negative_neutral_phrases.iterrows():
    sentence = s["Sentence"]
    width = 100
    print("-" * width)
    for i in range(0, len(sentence), width):
        print(sentence[i : i + width])
    print("-" * width)
    print()

----------------------------------------------------------------------------------------------------
overall i can't give it a good recommendation umm
----------------------------------------------------------------------------------------------------

----------------------------------------------------------------------------------------------------
but umm i really wouldn't recommend this
----------------------------------------------------------------------------------------------------

----------------------------------------------------------------------------------------------------
not to say that he wants to come to canada but if he ever does he would have a criminal record and t
he canadian authorities the border authorities the they're called canadian border service agency off
icers could potentially deny him from entering canada and he may have to apply for a special permit 
called the temporary resident permit or to be admitted so the consequences of being convicted of an

## 2.2 Sentiment to Label
Estrategia de arredondar a média dos anotadores e transformar em variável categórica

In [13]:
def sentiment_to_category(row):
    sentiment = {
        -3: "highly negative",
        -2: "negative",
        -1: "weakly negative",
        0: "neutral",
        1: "weakly positive",
        2: "positive",
        3: "highly positive",
    }

    numerical_sentiment = row
    numerical_sentiment_rounded = np.round(numerical_sentiment)
    return sentiment.get(numerical_sentiment_rounded)


df["Sentiment_Categorical"] = df["Sentiment"].apply(sentiment_to_category)
display(df)

Sentence  Sentiment  \
0      it is written that we should not be yoked with...  -1.000000   
1      if we are to be children of god we have to be ...  -0.333333   
2      this is a very difficult message because even ...  -1.333333   
3      and even in your own family or your own friend...  -0.333333   
4      but what i do know is if you ask for the holy ...   1.333333   
...                                                  ...        ...   
22132  a lot of the movie is just more awkward than i...  -2.000000   
22133  suggest going out there and finding save yourself  -2.000000   
22134  the money get yourself another nicholas cage m...  -2.666667   
22135                  i don't know like face off that's   0.000000   
22136     a good john woo nicholas cage movie watch that   1.000000   

       Happiness   Sadness     Anger      Fear   Disgust  Surprise  \
0       0.000000  0.333333  2.000000  0.000000  0.000000  0.333333   
1       0.000000  0.000000  0.333333  0.000000  0.666667  0.000000   
2       0.000000  0.333333  1.000000  0.333333  0.000000  0.000000   
3       0.000000  0.000000  0.000000  0.000000  0.000000  0.333333   
4       0.333333  0.000000  0.000000  0.000000  0.000000  0.000000   
...          ...       ...       ...       ...       ...       ...   
22132   0.000000  0.333333  1.000000  0.000000  1.666667  0.000000   
22133   0.000000  0.666667  0.333333  0.000000  0.333333  0.000000   
22134   0.000000  0.333333  0.666667  0.000000  0.666667  0.000000   
22135   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
22136   0.000000  0.000000  0.333333  0.000000  0.000000  0.000000   

      Emotion_Label Sentiment_Categorical  
0       Non-Neutral       weakly negative  
1       Non-Neutral               neutral  
2       Non-Neutral       weakly negative  
3       Non-Neutral               neutral  
4       Non-Neutral       weakly positive  
...             ...                   ...  
22132   Non-Neutral              negative  
22133   Non-Neutral              negative  
22134   Non-Neutral       highly negative  
22135       Neutral               neutral  
22136   Non-Neutral       weakly positive  

[22137 rows x 10 columns]

In [14]:
df.groupby(by=["Sentiment", "Sentiment_Categorical"])["Sentence"].count().reset_index()

Sentiment Sentiment_Categorical  Sentence
0   -3.000000       highly negative       124
1   -2.666667       highly negative       278
2   -2.333333              negative       409
3   -2.000000              negative       644
4   -1.666667              negative       752
5   -1.333333       weakly negative       823
6   -1.000000       weakly negative       887
7   -0.666667       weakly negative      1106
8   -0.500000               neutral         4
9   -0.333333               neutral      1513
10  -0.166667               neutral         3
11   0.000000               neutral      4542
12   0.166667               neutral        11
13   0.333333               neutral      2933
14   0.500000               neutral        11
15   0.666667       weakly positive      2460
16   0.833333       weakly positive         7
17   1.000000       weakly positive      1850
18   1.166667       weakly positive         3
19   1.333333       weakly positive      1389
20   1.500000              positive         8
21   1.666667              positive       994
22   1.833333              positive         2
23   2.000000              positive       728
24   2.333333              positive       393
25   2.500000              positive         1
26   2.666667       highly positive       202
27   3.000000       highly positive        60

## 2.3 Emotions

Este conjunto de dados contém múltiplas sentenças, cada uma classificada com base na média de avaliações fornecidas por três anotadores diferentes, utilizando uma escala Likert.

Escala Likert:
 - 0: Nenhuma evidência da emoção.
 - 1: Pouca evidência da emoção.
 - 2: Evidência da emoção.
 - 3: Forte evidência da emoção.
Cada sentença é representada por um vetor numérico, onde cada elemento corresponde a uma das seis emoções identificadas:

Exemplo de vetor de emoções:

Happiness
Sadness
Anger
Fear
Disgust
Surprise
Exemplo de sentença rotulada:

text = [0.0, 0.3333, 0.0, 1.6667, 0.0, 0.0]

Classificação dos Rótulos:
A tarefa agora é conceber uma metodologia para converter estes rótulos numéricos em uma única categoria representativa para cada frase. As possíveis categorias são:

 - UniqueLabel: Sentenças que contêm apenas uma única média no vetor.
 - MultiLabel: Sentenças que contêm duas ou mais médias diferentes no vetor.
 - EqualMultiLabel: Sentenças que contêm duas ou mais médias iguais no vetor.

A classificação adequada é crucial para análises subsequentes e para treinar modelos de machine learning que dependem de rótulos claros e distintos.

In [15]:
def multiple_label(row):
    emotions = ["Happiness", "Sadness", "Anger", "Fear", "Disgust", "Surprise"]
    multiple_label = [emotion for emotion in emotions if row[emotion] > 0]
    if len(multiple_label) > 1:
        return True
    return False


def multiple_label_equal_average(row):
    """
    Verifica se uma linha possui múltiplos rótulos com a mesma média.

    Parameters:
    row (dict): Uma linha do dataset, onde as chaves são rótulos de emoções e os valores são suas respectivas médias.

    Returns:
    bool: Retorna True se existirem múltiplos rótulos com a mesma média máxima, caso contrário, False.
    """

    emotions = ["Happiness", "Sadness", "Anger", "Fear", "Disgust", "Surprise"]

    # Extraindo as médias (ignorando rótulos com média 0)
    averages = [row[emotion] for emotion in emotions if row[emotion] > 0]

    # Verifica se existem múltiplos rótulos e se a maior média aparece mais de uma vez
    if len(averages) > 1 and averages.count(max(averages)) > 1:
        return True
    return False


df["Multiple_Label"] = df.apply(multiple_label, axis=1)
df["Equal_Average"] = df.apply(multiple_label_equal_average, axis=1)

multiple_label_df = df.loc[df["Multiple_Label"] == True]
unique_label_df = df.loc[df["Multiple_Label"] == False]
equal_average_df = df.loc[df["Equal_Average"] == True]

count_unique_label = unique_label_df.shape[0]
count_multiple_label = multiple_label_df.shape[0]
count_equal_average = unique_label_df.shape[0]
count_non_equal_average_multiple_labels = count_multiple_label - count_equal_average

display(df)

Sentence  Sentiment  \
0      it is written that we should not be yoked with...  -1.000000   
1      if we are to be children of god we have to be ...  -0.333333   
2      this is a very difficult message because even ...  -1.333333   
3      and even in your own family or your own friend...  -0.333333   
4      but what i do know is if you ask for the holy ...   1.333333   
...                                                  ...        ...   
22132  a lot of the movie is just more awkward than i...  -2.000000   
22133  suggest going out there and finding save yourself  -2.000000   
22134  the money get yourself another nicholas cage m...  -2.666667   
22135                  i don't know like face off that's   0.000000   
22136     a good john woo nicholas cage movie watch that   1.000000   

       Happiness   Sadness     Anger      Fear   Disgust  Surprise  \
0       0.000000  0.333333  2.000000  0.000000  0.000000  0.333333   
1       0.000000  0.000000  0.333333  0.000000  0.666667  0.000000   
2       0.000000  0.333333  1.000000  0.333333  0.000000  0.000000   
3       0.000000  0.000000  0.000000  0.000000  0.000000  0.333333   
4       0.333333  0.000000  0.000000  0.000000  0.000000  0.000000   
...          ...       ...       ...       ...       ...       ...   
22132   0.000000  0.333333  1.000000  0.000000  1.666667  0.000000   
22133   0.000000  0.666667  0.333333  0.000000  0.333333  0.000000   
22134   0.000000  0.333333  0.666667  0.000000  0.666667  0.000000   
22135   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
22136   0.000000  0.000000  0.333333  0.000000  0.000000  0.000000   

      Emotion_Label Sentiment_Categorical  Multiple_Label  Equal_Average  
0       Non-Neutral       weakly negative            True          False  
1       Non-Neutral               neutral            True          False  
2       Non-Neutral       weakly negative            True          False  
3       Non-Neutral               neutral           False          False  
4       Non-Neutral       weakly positive           False          False  
...             ...                   ...             ...            ...  
22132   Non-Neutral              negative            True          False  
22133   Non-Neutral              negative            True          False  
22134   Non-Neutral       highly negative            True           True  
22135       Neutral               neutral           False          False  
22136   Non-Neutral       weakly positive           False          False  

[22137 rows x 12 columns]

### 2.3.1 UniqueLabel

Example Unique Label

In [16]:
display(unique_label_df)

Sentence  Sentiment  \
3      and even in your own family or your own friend...  -0.333333   
4      but what i do know is if you ask for the holy ...   1.333333   
6      and if you want to be separate and truly a chi...   0.333333   
8      even christians like this verse because they t...  -0.333333   
10     they think that they can be doing all the same...   0.000000   
...                                                  ...        ...   
22123  and martial arts really like to emphasize dram...   1.000000   
22124  second you know what i mean like throwing a th...  -0.333333   
22125  the room is very dramatic when everyone is sho...   0.333333   
22135                  i don't know like face off that's   0.000000   
22136     a good john woo nicholas cage movie watch that   1.000000   

       Happiness   Sadness     Anger  Fear  Disgust  Surprise Emotion_Label  \
3       0.000000  0.000000  0.000000   0.0      0.0  0.333333   Non-Neutral   
4       0.333333  0.000000  0.000000   0.0      0.0  0.000000   Non-Neutral   
6       0.000000  0.333333  0.000000   0.0      0.0  0.000000   Non-Neutral   
8       0.000000  0.000000  1.333333   0.0      0.0  0.000000   Non-Neutral   
10      0.000000  0.000000  1.000000   0.0      0.0  0.000000   Non-Neutral   
...          ...       ...       ...   ...      ...       ...           ...   
22123   0.000000  0.000000  0.666667   0.0      0.0  0.000000   Non-Neutral   
22124   0.000000  0.000000  0.000000   0.0      0.0  0.000000       Neutral   
22125   0.000000  0.000000  0.666667   0.0      0.0  0.000000   Non-Neutral   
22135   0.000000  0.000000  0.000000   0.0      0.0  0.000000       Neutral   
22136   0.000000  0.000000  0.333333   0.0      0.0  0.000000   Non-Neutral   

      Sentiment_Categorical  Multiple_Label  Equal_Average  
3                   neutral           False          False  
4           weakly positive           False          False  
6                   neutral           False          False  
8                   neutral           False          False  
10                  neutral           False          False  
...                     ...             ...            ...  
22123       weakly positive           False          False  
22124               neutral           False          False  
22125               neutral           False          False  
22135               neutral           False          False  
22136       weakly positive           False          False  

[13838 rows x 12 columns]

Para sentenças que apresentam um único valor de média em seu vetor, a estratégia adotada será a de rotulá-las de acordo com a categoria correspondente a essa média específica.

### 2.3.2 MultiLabel

In [17]:
def multiple_label(row):
    emotions = ["Happiness", "Sadness", "Anger", "Fear", "Disgust", "Surprise"]
    multiple_label = [emotion for emotion in emotions if row[emotion] > 0]
    if len(multiple_label) > 1:
        return True
    return False


def multiple_label_equal_average(row):
    """
    Verifica se uma linha possui múltiplos rótulos com a mesma média.

    Parameters:
    row (dict): Uma linha do dataset, onde as chaves são rótulos de emoções e os valores são suas respectivas médias.

    Returns:
    bool: Retorna True se existirem múltiplos rótulos com a mesma média máxima, caso contrário, False.
    """

    emotions = ["Happiness", "Sadness", "Anger", "Fear", "Disgust", "Surprise"]

    # Extraindo as médias (ignorando rótulos com média 0)
    averages = [row[emotion] for emotion in emotions if row[emotion] > 0]

    # Verifica se existem múltiplos rótulos e se a maior média aparece mais de uma vez
    if len(averages) > 1 and averages.count(max(averages)) > 1:
        return True
    return False


df["Multiple_Label"] = df.apply(multiple_label, axis=1)
df["Equal_Average"] = df.apply(multiple_label_equal_average, axis=1)

multiple_label_df = df.loc[df["Multiple_Label"] == True].copy()
unique_label_df = df.loc[df["Multiple_Label"] == False].copy()
equal_average_df = df.loc[df["Equal_Average"] == True].copy()

count_unique_label = unique_label_df.shape[0]
count_multiple_label = multiple_label_df.shape[0]
count_equal_average = equal_average_df.shape[0]
count_non_equal_average_multiple_labels = count_multiple_label - count_equal_average

#### Example MultiLabel

In [18]:
display(multiple_label_df)

Sentence  Sentiment  \
0      it is written that we should not be yoked with...  -1.000000   
1      if we are to be children of god we have to be ...  -0.333333   
2      this is a very difficult message because even ...  -1.333333   
5      is a teacher that is not like an earthly teach...  -0.666667   
7      and he will teach you and he will show you all...   1.000000   
...                                                  ...        ...   
22130                        it at all the musical score  -2.333333   
22131                                        was no good  -2.000000   
22132  a lot of the movie is just more awkward than i...  -2.000000   
22133  suggest going out there and finding save yourself  -2.000000   
22134  the money get yourself another nicholas cage m...  -2.666667   

       Happiness   Sadness     Anger      Fear   Disgust  Surprise  \
0       0.000000  0.333333  2.000000  0.000000  0.000000  0.333333   
1       0.000000  0.000000  0.333333  0.000000  0.666667  0.000000   
2       0.000000  0.333333  1.000000  0.333333  0.000000  0.000000   
5       0.333333  0.000000  1.000000  0.000000  0.333333  0.000000   
7       0.666667  0.000000  0.666667  0.000000  0.000000  0.000000   
...          ...       ...       ...       ...       ...       ...   
22130   0.000000  0.000000  0.666667  0.000000  0.666667  0.000000   
22131   0.000000  0.000000  0.666667  0.000000  0.333333  0.000000   
22132   0.000000  0.333333  1.000000  0.000000  1.666667  0.000000   
22133   0.000000  0.666667  0.333333  0.000000  0.333333  0.000000   
22134   0.000000  0.333333  0.666667  0.000000  0.666667  0.000000   

      Emotion_Label Sentiment_Categorical  Multiple_Label  Equal_Average  
0       Non-Neutral       weakly negative            True          False  
1       Non-Neutral               neutral            True          False  
2       Non-Neutral       weakly negative            True          False  
5       Non-Neutral       weakly negative            True          False  
7       Non-Neutral       weakly positive            True           True  
...             ...                   ...             ...            ...  
22130   Non-Neutral              negative            True           True  
22131   Non-Neutral              negative            True          False  
22132   Non-Neutral              negative            True          False  
22133   Non-Neutral              negative            True          False  
22134   Non-Neutral       highly negative            True           True  

[8299 rows x 12 columns]

#### Example EqualMultiLabel

In [19]:
display(equal_average_df)

Sentence  Sentiment  \
7      and he will teach you and he will show you all...   1.000000   
13     i bet they're really devastated to see their k...  -1.666667   
46     but uhh you know even he couldn't save this movie  -1.666667   
48                                     plot was terrible  -1.333333   
62     when i was a kid it gave me bad dreams for wee...  -1.333333   
...                                                  ...        ...   
22114  this is the nicholas cage film directed by joh...   0.000000   
22116  i'm going to have to say do not get this movie...  -2.333333   
22121             john woo does martial arts he does not   0.000000   
22130                        it at all the musical score  -2.333333   
22134  the money get yourself another nicholas cage m...  -2.666667   

       Happiness   Sadness     Anger  Fear   Disgust  Surprise Emotion_Label  \
7       0.666667  0.000000  0.666667   0.0  0.000000  0.000000   Non-Neutral   
13      0.000000  0.333333  0.333333   0.0  0.333333  0.000000   Non-Neutral   
46      0.333333  0.333333  0.000000   0.0  0.333333  0.000000   Non-Neutral   
48      0.000000  0.333333  0.000000   0.0  0.333333  0.000000   Non-Neutral   
62      0.333333  0.000000  0.000000   0.0  0.333333  0.333333   Non-Neutral   
...          ...       ...       ...   ...       ...       ...           ...   
22114   0.666667  0.000000  0.666667   0.0  0.000000  0.000000   Non-Neutral   
22116   0.000000  0.333333  0.000000   0.0  0.333333  0.000000   Non-Neutral   
22121   0.000000  0.000000  0.333333   0.0  0.333333  0.000000   Non-Neutral   
22130   0.000000  0.000000  0.666667   0.0  0.666667  0.000000   Non-Neutral   
22134   0.000000  0.333333  0.666667   0.0  0.666667  0.000000   Non-Neutral   

      Sentiment_Categorical  Multiple_Label  Equal_Average  
7           weakly positive            True           True  
13                 negative            True           True  
46                 negative            True           True  
48          weakly negative            True           True  
62          weakly negative            True           True  
...                     ...             ...            ...  
22114               neutral            True           True  
22116              negative            True           True  
22121               neutral            True           True  
22130              negative            True           True  
22134       highly negative            True           True  

[2810 rows x 12 columns]

#### Quantificando

In [20]:
print(f"Dataset            -> {count_df}")
print(f"MultiLabel         -> {count_multiple_label}")
print(f"Percentage         -> {np.round(count_multiple_label / count_df, 2) * 100}%")
print()

print(f"MultiLabel         -> {count_multiple_label}")
print(f"NonEqualMultiLabel -> {count_non_equal_average_multiple_labels}")
print(f"EqualMultiLabel    -> {count_equal_average}")

print()


print(
    f"Percentage of Multiple Labels com médias iguais {np.round( count_equal_average / count_multiple_label , 2) * 100}%"
)
print(
    f"Percentual Multiple Labels sem médias iguais {np.round(count_non_equal_average_multiple_labels / count_multiple_label, 2) * 100 }%"
)
print()

print("Porcentagem do dataset", count_equal_average / count_df)
print(f"Porcentam do dataset {count_non_equal_average_multiple_labels / count_df}")

Dataset            -> 22137
MultiLabel         -> 8299
Percentage         -> 37.0%

MultiLabel         -> 8299
NonEqualMultiLabel -> 5489
EqualMultiLabel    -> 2810

Percentage of Multiple Labels com médias iguais 34.0%
Percentual Multiple Labels sem médias iguais 66.0%

Porcentagem do dataset 0.12693680263811719
Porcentam do dataset 0.24795591091837196


#### 2.3.2.1 EqualMultiLabel

Distribuition of Repetitive Averages

In [21]:
def count_non_zero_equal_averages(row):
    # Filtra as médias maiores que zero e conta quantas vezes cada média aparece
    return pd.Series(row[row > 0].values).value_counts().max()


# Aplicando a função em cada linha para contar o número máximo de médias iguais
equal_average_df["Max_Equal_Averages"] = equal_average_df[
    ["Happiness", "Sadness", "Anger", "Fear", "Disgust", "Surprise"]
].apply(count_non_zero_equal_averages, axis=1)

# Agrupando pelo número de máximas médias iguais e contando as sentenças
equal_average_count = (
    equal_average_df.groupby("Max_Equal_Averages").count()["Sentence"].reset_index()
)

# Renomeando as colunas para maior clareza
equal_average_count.columns = ["Number of Equal Averages", "Sentence Count"]

# Exibindo o resultado
display(equal_average_count)

Number of Equal Averages  Sentence Count
0                         2            2337
1                         3             404
2                         4              59
3                         5              10

Distribuition of Average

In [22]:
def get_combinations(row):
    # Filtra as médias maiores que zero
    non_zero_averages = row[row > 0]
    # Conta a frequência de cada média não nula
    counts = non_zero_averages.value_counts()

    # Se todas as médias são únicas, não haverá combinações
    if all(value == 1 for value in counts):
        return None

    # Encontra a maior média entre as mais comuns
    max_count = counts.max()
    most_common_averages = counts[counts == max_count].index
    most_common_average = max(most_common_averages)

    return most_common_average, max_count


# Aplicando a função em cada linha para obter a média mais comum e sua contagem
combination_results = equal_average_df[
    ["Happiness", "Sadness", "Anger", "Fear", "Disgust", "Surprise"]
].apply(get_combinations, axis=1)

# Contando a frequência de cada combinação
combination_counts = {}

for result in combination_results:
    if result is not None:
        average, count = result
        if (average, count) not in combination_counts:
            combination_counts[(average, count)] = 1
        else:
            combination_counts[(average, count)] += 1

# Convertendo o dicionário em uma lista de dicionários para criar um DataFrame
combination_data = [
    {"Average": average, "Combination Size": size, "Count": count}
    for (average, size), count in combination_counts.items()
]

# Convertendo em DataFrame
combination_df = pd.DataFrame(combination_data)

# Exibindo os resultados
display(combination_df.groupby(by=["Average", "Combination Size"]).sum())

Count
Average  Combination Size       
0.166667 2                     4
0.333333 2                  1331
         3                   314
         4                    52
         5                    10
0.666667 2                   719
         3                    80
         4                     7
1.000000 2                   214
         3                     8
1.333333 2                    38
1.666667 2                    21
         3                     1
2.000000 2                     7
         3                     1
2.333333 2                     2
2.666667 2                     1

# 3. Data cleaning

In [23]:
def emotions_to_category(row):
    emotions = ["Happiness", "Sadness", "Anger", "Fear", "Disgust", "Surprise"]

    # Criar uma lista de tuplas (emocao, média)
    emotions_average = [(emotion, row[emotion]) for emotion in emotions]

    # Verifica se todos as médias são zeros
    if all(average == 0 for _, average in emotions_average):
        return "Neutral"

    # Conta quantas tuplas têm a média diferente de zero
    count_non_zero = sum(average != 0 for _, average in emotions_average)

    # Verifica se apenas uma tupla têm média diferente de zero
    if count_non_zero == 1:
        for emotion, average in emotions_average:
            if average != 0:
                return emotion
    
    # MultiLabel: Escolhe a emoção com a maior média se houver mais de uma média diferente de zero
    if count_non_zero > 1:
        sorted_emotions = sorted(emotions_average, key=lambda x: x[1], reverse=True)
        # Verifica se há um empate na maior média
        if sorted_emotions[0][1] == sorted_emotions[1][1]:
            return "Indefinido"  # Ou pode retornar None
        return sorted_emotions[0][0]

    return None

df["Emotion_Category"] = df.apply(emotions_to_category, axis=1)
display(df)

Sentence  Sentiment  \
0      it is written that we should not be yoked with...  -1.000000   
1      if we are to be children of god we have to be ...  -0.333333   
2      this is a very difficult message because even ...  -1.333333   
3      and even in your own family or your own friend...  -0.333333   
4      but what i do know is if you ask for the holy ...   1.333333   
...                                                  ...        ...   
22132  a lot of the movie is just more awkward than i...  -2.000000   
22133  suggest going out there and finding save yourself  -2.000000   
22134  the money get yourself another nicholas cage m...  -2.666667   
22135                  i don't know like face off that's   0.000000   
22136     a good john woo nicholas cage movie watch that   1.000000   

       Happiness   Sadness     Anger      Fear   Disgust  Surprise  \
0       0.000000  0.333333  2.000000  0.000000  0.000000  0.333333   
1       0.000000  0.000000  0.333333  0.000000  0.666667  0.000000   
2       0.000000  0.333333  1.000000  0.333333  0.000000  0.000000   
3       0.000000  0.000000  0.000000  0.000000  0.000000  0.333333   
4       0.333333  0.000000  0.000000  0.000000  0.000000  0.000000   
...          ...       ...       ...       ...       ...       ...   
22132   0.000000  0.333333  1.000000  0.000000  1.666667  0.000000   
22133   0.000000  0.666667  0.333333  0.000000  0.333333  0.000000   
22134   0.000000  0.333333  0.666667  0.000000  0.666667  0.000000   
22135   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
22136   0.000000  0.000000  0.333333  0.000000  0.000000  0.000000   

      Emotion_Label Sentiment_Categorical  Multiple_Label  Equal_Average  \
0       Non-Neutral       weakly negative            True          False   
1       Non-Neutral               neutral            True          False   
2       Non-Neutral       weakly negative            True          False   
3       Non-Neutral               neutral           False          False   
4       Non-Neutral       weakly positive           False          False   
...             ...                   ...             ...            ...   
22132   Non-Neutral              negative            True          False   
22133   Non-Neutral              negative            True          False   
22134   Non-Neutral       highly negative            True           True   
22135       Neutral               neutral           False          False   
22136   Non-Neutral       weakly positive           False          False   

      Emotion_Category  
0                Anger  
1              Disgust  
2                Anger  
3             Surprise  
4            Happiness  
...                ...  
22132          Disgust  
22133          Sadness  
22134       Indefinido  
22135          Neutral  
22136            Anger  

[22137 rows x 13 columns]

In [30]:
cleaned_df = df.loc[:, ["Sentence", "Emotion_Category"]].copy()
display(cleaned_df)

Sentence Emotion_Category
0      it is written that we should not be yoked with...            Anger
1      if we are to be children of god we have to be ...          Disgust
2      this is a very difficult message because even ...            Anger
3      and even in your own family or your own friend...         Surprise
4      but what i do know is if you ask for the holy ...        Happiness
...                                                  ...              ...
22132  a lot of the movie is just more awkward than i...          Disgust
22133  suggest going out there and finding save yourself          Sadness
22134  the money get yourself another nicholas cage m...       Indefinido
22135                  i don't know like face off that's          Neutral
22136     a good john woo nicholas cage movie watch that            Anger

[22137 rows x 2 columns]

In [31]:
cleaned_df.loc[cleaned_df["Emotion_Category"] != "Indefinido"]

Sentence Emotion_Category
0      it is written that we should not be yoked with...            Anger
1      if we are to be children of god we have to be ...          Disgust
2      this is a very difficult message because even ...            Anger
3      and even in your own family or your own friend...         Surprise
4      but what i do know is if you ask for the holy ...        Happiness
...                                                  ...              ...
22131                                        was no good            Anger
22132  a lot of the movie is just more awkward than i...          Disgust
22133  suggest going out there and finding save yourself          Sadness
22135                  i don't know like face off that's          Neutral
22136     a good john woo nicholas cage movie watch that            Anger

[19327 rows x 2 columns]